proposals/gc/MVP.md

```diff
@@ -97,20 +105,187 @@ New abbreviations are introduced for reference types in binary and text format,
+    - iff `<deftype0> ok` and extends the context accordingly
+    - and `<deftype>* ok` under the extended context
+
+* a plain type definition is valid if its `subtype` is at its type index
+  - `<subtype> ok` and extends the context with `<subtype>`
+    - iff `<subtype> ok($t)` where `$t` is the next unused (i.e., current) type index
+
+* a recursive type definition is valid if its types are valid under the context containing all of them
+  - `rec <subtype>* ok` and extends the context with `<ctxtype>*`
+    - iff `<subtype>* ok($t)` under the extended context(!)
+    - where `$t` is the next unused (i.e., current) type index
+    - and `N = |<subtype>*|-1`
+    - and `<ctxtype>*  = (rec <subtype>*).0, ..., (rec <subtype>*).N`
+
+* a sequence of subtype's is valid of each of them is valid for their respective index
+  - `<subtype0> <subtype>* ok($t)`
+    - iff `<subtype0> ok($t)`
+    - and `<subtype>* ok($t+1)`
+
+* an individual subtype is valid if its definition is valid, matches every supertype, and no supertype has an index higher than its own
+  - `sub $t* <strtype> ok($t')`
+    - iff `<strtype> ok`
+    - and `(<strtype> <: expand($t))*`
+    - and `($t < $t')*`
+  - Note: the upper bound on the supertype indices ensures that subtyping hierarchies are never circular, because definitions need to be ordered.
+
+* as [before](https://github.com/WebAssembly/function-references/proposals/function-references/Overview.md#types), a strtype is valid if all the occurring value types are valid
+  - specifically, a concrete reference type `(ref $t)` is valid when `$t` is defined in the context
+
+Example: Consider two mutually recursive types:
+```
+(rec
+  (type $t1 (struct (field i32 (ref $t2))))
+  (type $t2 (struct (field i64 (ref $t1))))
+)
+```
+In the context, these will be recorded as:
+```
+$t1 = rect1t2.0
+$t2 = rect1t2.1
+
+where
+
+rect1t2 = (rec
+  (struct (field i32 (ref $t2)))
+  (struct (field i64 (ref $t1)))
+)
+```
+That is, the types are defined as projections from their respective recursion group, using their relative inner indices `0` and `1`.
+
+
+#### Equivalence
+
+Type equivalence, written `t == t'` here, is essentially defined inductively. All rules are simply the canonical congruences, with the exception of the rule for recursive types.
+
+For the purpose of defining recursive type equivalence, type indices are extended with a special form that distinguishes regular from recursive type uses.
+
+* `rec.<i>` is a new form of type index
+  - `typeidx ::= ... | rec.<i>`
+
+This form is only used during equivalence checking, to identify and represent "back edges" inside a recursive type. It is merely a technical device for formulating the rules and cannot appear in source code. It is introduced by the following auxiliary meta-function:
+
+* Rolling a context type produces an _iso-recursive_ representation of its underlying recursion group
+  - `tie($t)                    = tie_$t(<ctxtype>)`  iff `$t = <ctxtype>`
```

Comment:
Should this be "iff `C($t) = <ctxtype>`" like in the `unroll` and `expand` definitions?